ivadomed/transforms.py

```diff
@@ -42,148 +138,76 @@ def __call__(self, sample):
-            and the value their parameters.
-        requires_undo (bool): If True, does not include transforms which do not have an undo_transform
-            implemented yet.
-    Returns:
-        torchvision.transforms.Compose object.
+        wspace (float): Resolution along the first axis, in mm.
+        hspace (float): Resolution along the second axis, in mm.
+        interpolation_order (int): Order of spline interpolation. Set to 0 for label data. Default=2.
     """
-    list_transform = []
-    for transform in dict_transforms.keys():
-        parameters = dict_transforms[transform]
-
-        # call transfrom
-        transform_obj = globals()[transform](**parameters)
-
-        # check if undo_transform method is implemented
-        if requires_undo:
-            if hasattr(transform_obj, 'undo_transform'):
-                list_transform.append(transform_obj)
-            else:
-                print('{} transform not included since no undo_transform available for it.'.format(transform))
-        else:
-            list_transform.append(transform_obj)
-
-    return torchvision_transforms.Compose(list_transform)
-
-
-class Resample(IMEDTransform):
 
-    def __init__(self, wspace, hspace,
-                 interpolation=Image.BILINEAR,
-                 labeled=True):
+    def __init__(self, wspace, hspace, interpolation_order=2):
         self.hspace = hspace
         self.wspace = wspace
-        self.interpolation = interpolation
-        self.labeled = labeled
-
-    @staticmethod
-    def do_resample(list_data, new_shape, interpolation_mode):
-        list_data_out = []
-        for i, data in enumerate(list_data):
-            resampled_data = data.resize(new_shape,
-                                         resample=interpolation_mode)
-            list_data_out.append(resampled_data)
-        return list_data_out
-
-    def undo_transform(self, sample):
-        rdict = {}
+        self.interpolation_order = interpolation_order
 
+    @list_capable
+    def undo_transform(self, sample, metadata):
         # Get original data shape
-        hshape, wshape = sample['input_metadata']['data_shape']
-
-        # Input data
-        rdict['input'] = self.do_resample(list_data=sample["input"],
-                                          new_shape=(wshape, hshape),
-                                          interpolation_mode=self.interpolation)
-
-        # Prediction data
-        # Note: We use here self.interpolation instead of forcing Image.NEAREST
-        #       in order to ensure soft output
-        rdict['gt'] = self.do_resample(list_data=sample["gt"],
-                                       new_shape=(wshape, hshape),
-                                       interpolation_mode=self.interpolation)
+        hshape, wshape = metadata['data_shape']
 
-        # Update
-        sample.update(rdict)
-        return sample
+        # Undo resampling
+        data_out = resize(sample,
```

Comment:
For GT, the `resize` function returns a soft output, should we binarize after resampling or using a soft output as GT is preferable?

Reply:
Good point
Few thoughts:
- I will try order = 0 for GT
- There is other interpolations after resampling (elastic transform, rotation...etc.) --> I feel that if we binarise after each transform we may accumulate a fair amount of error along the way (versus binarize at the very end of the transform pipeline)